syntax = "proto3";


enum Result {
  OK = 0;
  ID_NOT_EXIST = 1;
  DESKSERVER_OFFLINE = 2;
  RELAYSERVER_OFFLINE = 3;
  INNER_ERROR = 4;
}

message RequestRelay {
    enum DeskRole {
    DESK_CONTROL = 0;
    DESK_SERVER = 1;
    }
  string uuid = 1;
  DeskRole role = 2; 
}

message RegisterPeer {
  bytes uuid = 1;
}

message RegisterPeerResponse {
  Result result = 1;
}

message PunchHoleRequest { 
   bytes uuid = 1;
   bytes id = 2;
}

message PunchHoleResponse {
  string relay_server = 1;
  int32 relay_port = 2;
  Result result = 3;
}
message Heartbeat {}

message PunchHole { 
     bytes id = 1;
}

message PunchHoleSent {
  string relay_server = 1;
  int32 relay_port = 2;
  Result result = 3;
  bytes id = 4;
}

message InputControlEvent {
  oneof event {
    MouseEvent mouse_event = 1;
    KeyboardEvent keyboard_event = 2;
  }
}


message InpuVideoFrame{
  bytes data = 1;
 
}

message MouseEvent {
  int32 mask = 1;
  sint32 x = 2;
  sint32 y = 3;
  repeated ControlKey modifiers = 4;
}

message KeyboardEvent {
  ControlKey key = 2;
  bool pressed = 3;
}


message RendezvousMessage {
  oneof union {
    RequestRelay request_relay = 1;
    RegisterPeer register_peer = 2;
    RegisterPeerResponse register_peer_response = 3;
    PunchHoleRequest punch_hole_request = 4;
    PunchHoleResponse punch_hole_response = 5;
    Heartbeat heartbeat = 6;
    PunchHole punch_hole = 7;
    PunchHoleSent punch_hole_sent = 8;
    InpuVideoFrame inpuVideoFrame = 9;
    InputControlEvent inputControlEvent = 10;

  }
}

enum ControlKey {
  KEY_UNKNOWN = 0;
  KEY_ALT = 1;
  KEY_BACKSPACE = 2;
  KEY_CAPSLOCK = 3;
  KEY_CONTROL = 4;
  KEY_DELETE = 5;
  KEY_DOWN_ARROW = 6;
  KEY_END = 7;
  KEY_ESCAPE = 8;
  KEY_F1 = 9;
  KEY_F2 = 10;
  KEY_F3 = 11;
  KEY_F4 = 12;
  KEY_F5 = 13;
  KEY_F6 = 14;
  KEY_F7 = 15;
  KEY_F8 = 16;
  KEY_F9 = 17;
  KEY_F10 = 18;
  KEY_F11 = 19;
  KEY_F12 = 20;
  KEY_HOME = 21;
  KEY_LEFT_ARROW = 22;
  KEY_META = 23;
  KEY_PAGE_DOWN = 24;
  KEY_PAGE_UP = 25;
  KEY_RETURN = 26;
  KEY_RIGHT_ARROW = 27;
  KEY_SHIFT = 28;
  KEY_SPACE = 29;
  KEY_TAB = 30;
  KEY_UP_ARROW = 31;
  KEY_INSERT = 32;
}